### ISSUE TITLE
Scheduled jobs don't get executed if super_fetch! is enabled

### ISSUE THREAD

**Issue by nasa42** [mperham/sidekiq#3345]
Ruby version: 2.3.1
Sidekiq / Pro / Enterprise version(s): sidekiq 4.2.7, sidekiq-pro 3.4.1

Since we enabled `super_fetch!`, some of scheduled jobs do not run and our logs are filled with following error:

```
Connection timed out
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/connection/ruby.rb:111:in `rescue in _write_to_socket'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/connection/ruby.rb:104:in `_write_to_socket'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/connection/ruby.rb:131:in `block in write'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/connection/ruby.rb:130:in `loop'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/connection/ruby.rb:130:in `write'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/connection/ruby.rb:374:in `write'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/client.rb:271:in `block in write'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/client.rb:250:in `io'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/client.rb:269:in `write'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/client.rb:228:in `block (3 levels) in process'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/client.rb:222:in `each'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/client.rb:222:in `block (2 levels) in process'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/client.rb:367:in `ensure_connected'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/client.rb:221:in `block in process'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/client.rb:306:in `logging'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/client.rb:220:in `process'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis/client.rb:120:in `call'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis.rb:1576:in `block in zrem'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis.rb:58:in `block in synchronize'
/usr/lib/ruby/2.3.0/monitor.rb:214:in `mon_synchronize'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis.rb:58:in `synchronize'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/redis-3.3.2/lib/redis.rb:1575:in `zrem'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.2.7/lib/sidekiq/scheduled.rb:25:in `block (2 levels) in enqueue_jobs'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.2.7/lib/sidekiq/scheduled.rb:15:in `each'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.2.7/lib/sidekiq/scheduled.rb:15:in `block in enqueue_jobs'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.2.7/lib/sidekiq.rb:95:in `block in redis'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/connection_pool-2.2.1/lib/connection_pool.rb:64:in `block (2 levels) in with'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/connection_pool-2.2.1/lib/connection_pool.rb:63:in `handle_interrupt'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/connection_pool-2.2.1/lib/connection_pool.rb:63:in `block in with'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/connection_pool-2.2.1/lib/connection_pool.rb:60:in `handle_interrupt'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/connection_pool-2.2.1/lib/connection_pool.rb:60:in `with'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.2.7/lib/sidekiq.rb:92:in `redis'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.2.7/lib/sidekiq/scheduled.rb:14:in `enqueue_jobs'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.2.7/lib/sidekiq/scheduled.rb:77:in `enqueue'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.2.7/lib/sidekiq/scheduled.rb:68:in `block in start'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.2.7/lib/sidekiq/util.rb:17:in `watchdog'
/home/webapps/myapp/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.2.7/lib/sidekiq/util.rb:25:in `block in safe_thread' 
```

This seems to happen only when `super_fetch!` is enabled. We even increased `network_timeout` to 5 but that didn't seem to help either. Also the payload of these jobs is very small (just a single integer).

Initializer:

``` ruby
Sidekiq.default_worker_options = {
  backtrace: true,
  retry: 1
}

#################################################
# Begin Sidekiq Status
#################################################

Sidekiq.configure_client do |config|
  config.client_middleware do |chain|
    chain.add Sidekiq::Status::ClientMiddleware, expiration: 1.hour
  end
end

Sidekiq.configure_server do |config|
  config.server_middleware do |chain|
    chain.add Sidekiq::Status::ServerMiddleware, expiration: 1.hour
  end
  config.client_middleware do |chain|
    chain.add Sidekiq::Status::ClientMiddleware, expiration: 1.hour
  end
end

#################################################
# End Sidekiq Status
#################################################


#################################################
# Begin Sidekiq Pro
#################################################

Sidekiq.configure_server do |config|
  config.super_fetch!
end

#################################################
# End Sidekiq Pro
#################################################


Sidekiq.configure_server do |config|
  config.server_middleware do |chain|
    # reset thread-local variables for Bugsnag as Sidekiq reuses threads between jobs
    # and that may leak thread-local variables between jobs.
    chain.add SidekiqMiddlewareResetBugsnag
  end
end

#################################################
# Feed Datadog
#################################################

Sidekiq.configure_server do |config|
  config.server_middleware do |chain|
    chain.add Sidekiq::Middleware::Server::Datadog
  end
end

#################################################
# End Datadog
#################################################
```

**Comment by mperham** [comment]
Is it possible your scheduled job queue is enormous?  How big is "Scheduled" in the Web UI?

**Comment by nasa42** [comment]
It spans across 26 pages now (25 per pages). We schedule them to be executed `30.seconds` but seems like none of them being executed since we enabled `super_fetch!`.

**Comment by mperham** [comment]
It shouldn't take 1 sec to zrem on a scheduled queue with less than 1000 items.  Something might be wrong with your Redis instance.  You can try activating the reliable scheduler and increase the redis client timeout:

```ruby
Sidekiq.configure_server do |config|
  config.redis = { network_timeout: 5 }
end
```

**Comment by nasa42** [comment]
As mentioned above, we already have `network_timeout` set to `5`.

**Comment by mperham** [comment]
I have no idea what's wrong then.

**Comment by nasa42** [comment]
We'll see if we can find more info.

**Comment by nasa42** [comment]
The problem was that the Sidekiq scheduler works on both `retry` and `schedule` set (https://github.com/mperham/sidekiq/blob/v4.2.7/lib/sidekiq/scheduled.rb#L8) and in the `retry` set we had an old job which had quite a large payload and was the reason of timeout-errors. Clearing out our Retry list solved the issue and all scheduled jobs got processed.

This issue had nothing to do with `super_fetch!` (except for the fact that that old job appeared again in the queue after we enabled `super_fetch!`. My hypothesis is that since `super_fetch!` manages its own private queue, that old job got stuck in that private queue when we disabled `super_fetch!` last month, and it got revived when we enabled `super_fetch!` again recently).

Closing.
